crates/svm-runtime-c-api/tests/api_tests.rs

```diff
@@ -1,23 +1,148 @@
+#![allow(unused)]
+
 extern crate svm_runtime_c_api;
 
 use svm_runtime_c_api as api;
-use svm_runtime_c_api::svm_byte_array;
 
+use std::convert::TryFrom;
 use std::ffi::c_void;
 
-use svm_abi_encoder::Encoder;
 use svm_codec::api::raw;
+use svm_ffi::{svm_byte_array, svm_env_t};
 use svm_layout::DataLayout;
-use svm_runtime::testing::WasmFile;
-use svm_types::{Address, State};
+use svm_runtime::{testing::WasmFile, vmcalls, Context};
+use svm_types::{Address, State, WasmType, WasmValue};
+
+use svm_sdk::traits::Encoder;
+use svm_sdk::ReturnData;
+
+/// We should land here when `trampoline` has been called with `host_env` containing
+/// a function index equaling to `COUNTER_MUL_FN_INDEX`
+fn counter_mul(ctx: &mut Context, args: &[WasmValue]) -> Result<Vec<WasmValue>, &'static str> {
+    assert_eq!(args.len(), 2);
+
+    let var_id = args[0].as_i32().unwrap();
+    let mul = args[1].as_i32().unwrap();
+
+    let old = vmcalls::get32(ctx, var_id);
+    let new = old * mul;
+
+    let results = vec![WasmValue::I32(new)];
+
+    Ok(results)
+}
+
+/// This struct will serve as our `host_env`.
+/// We use here integers as function identifiers to be called by the `trampoline`.
+#[derive(Debug)]
+#[repr(C)]
+struct func_index_t(u32);
+
+type Callback = fn(&mut Context, &[WasmValue]) -> Result<Vec<WasmValue>, &'static str>;
+
+const COUNTER_MUL_FN_INDEX: u32 = 123;
+
+/// Given a function identifier stored as part of the `host_env`
+/// we can know what Rust native function to call
+fn func_index_to_callback(func_idx: &func_index_t) -> Option<Callback> {
+    match func_idx.0 {
+        COUNTER_MUL_FN_INDEX => Some(counter_mul),
+        _ => None,
+    }
+}
+
+unsafe fn prepare_args(args: *const svm_byte_array) -> Result<Vec<WasmValue>, &'static str> {
+    let args: &svm_byte_array = &*args;
+
+    Vec::<WasmValue>::try_from(args).map_err(|_| "Invalid args")
+}
+
+fn wasm_error(msg: String) -> *mut svm_byte_array {
```

Comment:
Any reason for still using this instead of `svm_wasm_error_create`?

Reply:
`svm_wasm_error_create` expects `svm_byte_array`.

I can make it work but I prefer to keep the current code as it is.